### PR TITLE
[Feature] Added tests for token functionalities

### PR DESCRIPTION
**Pull request by nup9151f**
## Summary

Following #57 Adding more tests for token functionalities.

## Description

Changes:

- Added tests for event handlers in `ERC20Base`, `ERC721Base`, `ERC721LazyMint`

## Related Issue/Bounty

Continuation of #57

### CODE REVIEW

**Review by george-openformat — CHANGES_REQUESTED** [review]
This is looking so good, and fantastic it has caught some bugs.

Mainly just two things
- refactor setting up context into `beforeEach`
- be explicit in the event parameters for each test

File: tests/core/ERC20Base.test.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        createMockedFunction(Address.fromString(TEST_TOKEN_ID), "balanceOf", "balanceOf(address):(uint256)")
+            .withArgs([ ethereum.Value.fromAddress( receiver ) ])
+            .returns([ ethereum.Value.fromUnsignedBigInt(balanceReceiver) ]);
+
+        createMockedFunction(Address.fromString(TEST_TOKEN_ID), "totalSupply", "totalSupply():(uint256)")
+            .withArgs([])
+            .returns([ ethereum.Value.fromUnsignedBigInt(totalSupply) ]);
+
+        const token = getTestFungibleToken();
+        token.save();
```
[the remark this answers is not in the document]
Comment (nup9151f) [reply tests/core/ERC20Base.test.ts:37]: We don't need to use `createERC20Token`.
All functionalities in `ERC20Base` just need that the specified token exists in the store. We do not need to create events and trigger handlers to accomplish this, we can create the token directly.
The create erc20 token flow is tested in the facet tests.

File: tests/core/ERC721LazyMint.test.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            .withArgs([ ])
+            .returns([ ethereum.Value.fromUnsignedBigInt(BigInt.fromString(TEST_BADGETOKEN_ID).plus(One)) ]);
+
+        const badgeToken = getTestBadgeTokenEntity(
+            Address.fromString(TEST_TOKEN_ID), 
+            BigInt.fromString(TEST_BADGETOKEN_ID),
+            TEST_USER2_ID,
+            TEST_TOKEN_MINTED_URI
+        );
+        badgeToken.save();
```
[the remark this answers is not in the document]
Comment (nup9151f) [reply tests/core/ERC721LazyMint.test.ts:71]: This is the same test than `"Token minted new"` but with a BadgeToken that already exists in the store.
I am not sure whether the badge token already exists when this event handler is triggered.

File: tests/core/ERC20Base.test.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+        transferERC20();
```
[the remark this answers is not in the document]
Comment (nup9151f) [reply tests/core/ERC20Base.test.ts:91]: This is an error.
I changed `transferERC20` to receive `from` and `to` as parameters, now burnt supply updates correctly.

File: tests/core/ERC20Base.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (george-openformat) [reply tests/core/ERC20Base.test.ts:39]: Ah interesting thank you for clarifying!

Can't help but feeling that level of coupling for `ERC721` isn't the correct approach but that's an issue for another day!

File: tests/core/ERC721LazyMint.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (george-openformat) [reply tests/core/ERC721LazyMint.test.ts:69]: Gotcha! I miss interpreted "existing" for `badge` entity and not `badgeToken` entity. This makes total sense when the token has been lazy minted. 

Can the owner be set to `null`?  Looking at the[docs](https://www.assemblyscript.org/types.html#nullability) this is probably more tricky than it is worth.

**Review by george-openformat — APPROVED** [review]
Thanks for those changes and clarifying

If making existing badgeToken owner = null takes more than 5 mins then what you have is perfect.

**Review by tinypell3ts — APPROVED** [review]

File: tests/core/ERC721LazyMint.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nup9151f) [reply tests/core/ERC721LazyMint.test.ts]: > Can the owner be set to `null`? Looking at the[docs](https://www.assemblyscript.org/types.html#nullability) this is probably more tricky than it is worth.

No problem :+1: 
I also removed `core` folder and moved tests up to mirror `src` folder.

**Pull request merged**